csrc/multidevice/pipeline.h

```diff
@@ -0,0 +1,142 @@
+#pragma once
+#include <fusion.h>
+#include <ir/base_nodes.h>
+#include <multidevice/device_mesh.h>
+#include <multidevice/multidevice.h>
+#include <multidevice/pipeline_ir.h>
+
+/*
+This file implements the Pipeline interface.
+A Pipeline represents a Fusion segmented into a series of Stages, each Stage
+being thought as a portion of the original Fusion that will be executed on a
+given device mesh.
+
+The decomposition of the Pipeline into Stages is described through a
+PipelineDescriptor, which is nothing but a vector of PipelineStageDescriptor.
+Each PipelineStageDescriptor desribes a stage by listing all the
+Vals that compose it.
+
+During initialization, the system will figure out the data dependencies and
+infer what is each stage's inputs and outputs, defined as follow: a Val is a
+stage's input (resp. output) if it is either a global input (resp. output) of
+the originalFusion or if it is a direct consumer (resp. producer) of a Val from
+a another stage.
+
+To be considered as valid, the PipelineDescriptor must ensure that the described
+Pipeline satisfies the following assumption:
+*) Each Val belongs to one and only one Stage
+*) The directed graph composed of the Stages and their dependencies is acyclic
+   (i.e., the Pipeline is a DAG (Directed Acyclic Graph))
+   Remark: the word "Pipeline" can be misleading because it conveys the idea of
+a linear dependency between stages -- however our Pipeline is allowed to have
+any DAG structure
+*) The inputs of a stage are TensorViews which are either global inputs of the
+fusion OR are defined as copies (a "set" operation) of a Tv from another stage
+    (Later we could add the case where they are defined as a reduction).
+*) Global inputs of the Fusion belong to stage(s) whose mesh contains only one
+device index. Note: Later, when we add a new parallel type for inter-device
+sharding (dIdx, dIdy, etc...) we can loosen this condition by assuming only that
+the input values are not replicated on different devices. This assumption is
+natural because the input values must be univoque
+
+The PipelineStageDescriptor are passed to the Pipeline (through a
+PipelineDescriptor) by REFERENCE. As a consequence, after instantiation of the
+Pipeline, it is not allowed to add Vals to the PipelineStageDescriptor nor to
+free them. However, even after the Pipeline is instantiated, the user can still
+set som runtime informations on the PipelineStageDescriptor, which are
+- auto_schedule: bool indicating whether the stage should be autoscheduled
+- mesh (see multidevice/device_mesh.h) which is basically an n-array of devices
+indices on which the stage should be executed at runtime.
+*/
+
+namespace nvfuser {
+
+// Interface to describe the composition of a PipelineStage
+class TORCH_CUDA_CU_API PipelineStageDescriptor final {
+ public:
+  PipelineStageDescriptor() : unique_id(running_unique_id_++) {}
+
+  // The mesh on which the stage will be executed at runtime.
+  DeviceMesh mesh;
```

Comment:
I think this is execution information, can we pull this out of the descriptor so that the descriptors only contain information about how the DAG is partitioned, and the executor own the information on which stage is assigned to which device mesh?

Reply:
You are right, btw so is the field `auto_schedule`. Initially those two were actually wrapped in a struct called RuntimeInfo, that I removed for readability (but I could revert this change)
Those runtime info are neither written nor read by the pipeline constructor; they are simply carried untouched to be used later. You can even see in the test (test/test_gpu_multidevice.cpp:200) that those runtime info can be set after the pipeline creation.

The only reason why I think we should keep it this way is just for the user's convenience.
Indeed, it is easy to set those runtime params along the definition of the pipeline; on the other hand, If I define the pipeline, how can I later on easily access each stage to set the runtime info? The only solution is to keep track of pointers to each stage and to have some kind of map held by the pipeline.
I you think of a better design I would be happy to hear.

Even in the general case of pipelining for task-parallelism or other purposes, we could imagine that stages are able to hold some parameters irrelevant to the pipeline itself but used later. Do you think it is an issue?

I would just like to mention that, if Im not mistaken, this type of issue can be found in other places in nvFuser, e.g., setting the parallel type (which is not so different than setting the device mesh) is runtime info but it is set during Fusion definition

Reply:
I think I am OK with having `PipelineStage` storing multidevice runtime info, or other scheduling info for task parallelism. But that info should not be hard coded for one specific application. So I don't want `PipelineStageDescriptor` to have device mesh and `auto_schedule` there. One easy refactor we can do is:
```C++
// in ir/ directory
class PipelineStageDescriptor : public PolymorphicBase {};

// in multidevice/ directory
class MultiDeviceStageDescriptor : public PipelineStageDescriptor {
  DeviceMesh mesh;
  ...
};
```

I believe it would be helpful to realize that we are actually working on two independent projects, instead of one project:
- Project 1: Design and implement an interface to partition a DAG into subgraphs with different level of granularity, without assuming how these partition will be used. Build out utilities for operating on these partitions.
- Project 2: Design and implement multidevice support using the product of Project 1.

Project 1 should go to `ir/` or just `csrc/` directory. Project 2 should go to `multidevice/` directory. We should push forward project 1 prioritizing its only current customer: project 2. But project 1 is a separate project, it should be designed as a general approach for various applications.

Reply:
Thanks for your comment. Yes, what you suggest is very clear and is exactly what I have in mind. I would just prefer to merge this PR and move from there. First step would be indeed to clearly separate the pipelining logic from multidevice, also to make the pipelining composable, and besides I would also like to advance on other fronts of the multidevice support. That's why I think this would be easier to merge this pr and then move in small steps. Do you agree or do you want those change to be part of this pr?

Reply:
Yes, it totally makes sense to leave this to next PR.